src/shell/shell_cmds.cpp

```diff
@@ -2572,3 +2573,121 @@ void DOS_Shell::CMD_MOVE(char* args)
 		}
 	}
 }
+
+static std::vector<std::string> search_files(std::string_view query)
+{
+	const auto save_dta = dos.dta();
+	dos.dta(dos.tables.tempdta);
+	const auto dta = DOS_DTA(dos.dta());
+
+	bool found = DOS_FindFirst(std::string(query).c_str(),
+	                           ~(DOS_ATTR_VOLUME | DOS_ATTR_DIRECTORY));
+
+	std::vector<std::string> files = {};
+	while (found) {
+		DOS_DTA::Result result = {};
+		dta.GetResult(result);
+		files.emplace_back(std::move(result.name));
+		found = DOS_FindNext();
+	}
+
+	dos.dta(save_dta);
+	
+	return files;
+}
+
+void DOS_Shell::CMD_FOR(char* args)
+{
+	HELP("FOR");
+	
+	static const auto delimiters = std::string_view(",;= \t");
+
+	auto argsview = std::string_view(args);
+	auto consume_next_token = [&argsview]() -> std::optional<std::string_view> {
+
+		const auto start = argsview.find_first_not_of(delimiters);
+		if (start >= argsview.size()) {
+			return {};
+		}
+
+		const auto end = argsview.substr(start).find_first_of(delimiters);
+		if (end >= argsview.size()) {
+			return {};
+		}
+
+		const auto token = argsview.substr(start, end);
+		argsview         = argsview.substr(end + 1);
+		return token;
+	};
+
+	const auto variable = consume_next_token();
+	if (!variable || variable->size() != sizeof('%') + 1 ||
+	    variable->front() != '%') {
+		return;
+	}
+
+	const auto in_keyword = consume_next_token();
+	if (!in_keyword || !iequals(*in_keyword, "IN")) {
+		return;
+	}
+
+	const auto parameters =
+	        [&argsview]() -> std::optional<std::vector<std::string>> {
+		const auto list_start = argsview.find_first_of('(');
+		if (list_start >= argsview.size()) {
+			return {};
+		}
+		const auto list_end = argsview.substr(list_start).find_first_of(')');
+		if (list_end >= argsview.size()) {
+			return {};
+		}
+		auto string_list = split(argsview.substr(list_start + sizeof('('),
+		                                         list_end - sizeof(')')),
+		                         delimiters);
+		argsview = argsview.substr(list_end + sizeof(')') + 1);
+
+		auto parameters = std::vector<std::string>();
+		for (auto& str : string_list) {
+			if (str.find('*') == std::string::npos &&
+			    str.find('?') == std::string::npos) {
+				parameters.emplace_back(std::move(str));
+				continue;
+			}
+			auto files = search_files(str);
+			parameters.insert(parameters.end(),
+			                  std::make_move_iterator(files.begin()),
+			                  std::make_move_iterator(files.end()));
+		}
+		return parameters;
```

Comment:
Can you drop this whole `string_view` and  move business, please; we're parsing an up to 256 char max string here, not a 100 MB XML 😅 All this is unnecessary complication at the code level.

Please reimplement this in the most basic and primitive way that does the job. If we're creating temporary strings left, right, center, who cares — this is not performance-intensive code.

Reply:
(Otherwise, the code is nice and clearly structured, for 🥇 on that. But let's keep it simple 😅)

Reply:
1.6x slowdown, 10x slowdown, or 100x slowdown is meaningless without context.

I'm advocating for using a simple style and only complicating things when they make a tangible benefit. Getting better results in micro-benchmarks is not a tangible benefit. Overall, observable _whole-application performance_ and behaviour is.

Even in the DOS days, well, even before that, in the 80s, people realised such exercises are fairly meaningless in isolation. People wrote their games in C, then rewrote the hot loops in asm after benchmarking, and that's it. Config parsing, disk I/O—just use C, performance is a non-issue there.

I get what you're saying, though; when everything is slow, nothing really jumps out when benchmarking the top 100 slowest functions. But as I see it, we could write the shell commands in Python, and even them being 1000x slower than the "worst C++ code" you can imagine, no one would notice a difference even on the slowest hardware.

> Again, it's not about this one function in particular, it's about how it spreads to the program as a whole.

Yes, so it does not spread, because this is a FOR loop you execute from the DOS prompt... If you were poking around in the CPU emulation or something similar, yeah, then we need to be careful about such things. Not for DOS commands though that run once in a blue moon.

Reply:
I'm not sure I'm seeing where or how the code is being harmed in exchange for the micro optimizations.

In other words: can it be written in much fewer and easier to read lines, if we're willing to forfeit the measured 1.6x performance boost of string_view vs string copies? 

I don't see that a trade like that is possible (if I did, I would offer it as a code replacement suggestion). 

Yeah: I do see move being used, but it tells me wer're just taking a rock from pile A and placing it on pile B. 

Could the emplace, insert, move-iterator-makers being replaced with a for loop? Yup, it would be about the samr number of lines and would probably be a bit less busy to look at.

Reply:
> Could the emplace, insert, move-iterator-makers be replaced with a for loop? Yup, it would be about the same number of lines and would probably be a bit less busy. So maybe that's a small but minor suggestion.

That's a big part of it, I think; move and string_view just add insult to injury 😅 Maybe it would be quite alright by introducing a few helpers to string utils to remove all that iterator noise.

Reply:
Also, can't we just take the incoming string, split it by the delimiters, that gives us a vector of strings. Then just operate on that vector, so need for that tokeniser that returns string views, etc. Just make it simpler. We're only dealing with less than 1000 bytes of input here... You could copy that 10 times and it still wouldn't matter one bit. Whatever is more readable and hopefully all that iterator business would go away.

Reply:
> As a contributor, I hope you can at least appreciate how frustrating it can be to just be coding using completely run-of-the-mill C++, and to be told it's micro-optimization and that it complicates the code, without any further elaboration.

I don't think "run-of-the-mill C++" exists. For example, many people don't use the std lib at all—that's still C++. Many people still use it as "C with classes" and none of the C++11 and later features–that's still C++.

I'm hearing that you're frustrated, but then these discussions happen regularly only when reviewing your PRs @MeAreJeenius, not with other contributor's PRs. In a way, it's frustrating for me too, because if I contribute to another project, then I try to follow their established standards not change their project. I _might_ bring up my ideas for improvement once or twice, but if they don't want, I will not push it further.

Because you've asked about it, here's my personal perspective:

Many people are not that familiar with `string_view` and `move`, I'm actually one of them. I don't like C++, I don't use it daily at work (and that's the last thing I want to do 😅), I just want to use a minimal subset here that does the job. To put it differently, my love for DOSBox is greater than my dislike for C++ 😄

It's judgement call, but I personally think there are lots of gotchas and complications with `std::move` and `string_view`, you need to be careful to use them. E.g. if you store a string view somewhere, then the string it points to gets freed, well you're in trouble—that's extra complexity. Then I wanted to learn more about `std::move` and found blog posts that go on for 5-10-20 pages, discussing all the intricate details, edge case, gotchas, etc. No thanks, we can solve everything with pointers and references. There are already too many ways to shoot yourself in the foot in C++, these just give you more rope to hang yourself, metaphorically speaking.

From my experience, the problem with languages that have too many features is that people start using _all_ those features. Then if someone new want to contribute, he/she has to be familiar with all the 10 different ways of doing things, when in reality just restricting ourselves to 2 or 3 ways would have sufficed.

So basically that's where I'm coming from. I like to do the simplest solution that does the job.

Reply:
@MeAreJeenius , the way you've described these use cases is clear to me, and when re-reading your code, as you've explained, actually makes these aspects more self-documenting:

#### string_view

- I want to see a char buffer -> I use string_view. Simple. No need to think about why to use it or why it's there.
- Can't accidentally change characters if you forget const

#### move

- I want this object to go over there -> I use move. Simple. No need to think about why to use it or why it's there.
- Accidental use after move is caught by CI / SA. Accidental use after copy is not.

So I see your use as justified, especially when the intent is simple and clear as you've described. 

I think the dillema as reviewers is we haven't had a lot of exposure to these (so John and I might wonder.. is this just overly faffy syntax bloat?). But perhaps with fresh eyes, the intent becomes much clearer.

Things do get muddy when  lamdas are put inline into  algorithms, combined with iterator generators and others. I find sometimes going back to for loops easier to see. 

Would you agree with that take, @johnnovak?

Reply:
Yeah pretty much @kcgen, that's a good summary. I think there's a lot going on there that makes me wonder, is this really the simplest way to do this relatively simple job? But the C++ is very syntax heavy still, we could improve that with some helper functions but it's still far from pseudo-code level cleanness.

Reply:
Good point about string_view introducting new concerns about lifetime of the content versus the object itself. The two are no longer coupled. 

There are probably some good rules of thumb or use-cases that can make this easy to review and ensure won't bite us: like the string_view objects only live (or have scope) as long as their content is in-scope. 

So self-contained use inside a function means that the string_view cannot outlive its content.

@MeAreJeenius , are there some rules like that?

Reply:
To be fair, using string views only within a single function is as safe as it gets, no doubt. I guess I was reacting more to the _sum_ of all these things that are going on here, as @kcgen put it 😅 I'm a simple man, and I'm getting old, I just want it simple, hey 😎

Reply:
My other problem with string views is that `.data()` is not guaranteed to return null-terminated substrings, right? That's actually hassle and another source for bugs (e.g. if you pass in std::strings as string views to some function, things will work fine when just requesting `data()` then passing that to C-style APIs... until you start passing in _real_ substrings, not full strings, then things will start breaking).

Reply:
For the record, I did not say to stop using lambdas. I use them all the time, I too like to keep one-off helpers within the scope (unless when they grow too big).

> I'm not enjoying contributing anymore, since it's clear what I want out of it, and what the maintainers want are at odds.

That's a personal choice and entirely up to you, but I agree that reviewing relatively simple changes like this should not feel like a constant fight. Ultimately, contributors to any project are kind of expected to adapt/adjust to the style of the project. Improvements over time can be made, of course, but usually over time.

Reply:
@MeAreJeenius , I'm not sure if you've been following and reading all the other team's PRs that have been going into the code base. 

As an example, you'll see that we never use `type var{};` initializers, so we were off to a rough start.
 - One one hand, the C++ committee claimed this is the more efficient initialzer, so please migrate from ` = value` :facepalm: 
 - On the other hand, compilers are free to eliminate unnecessary copies: so we see counter-arguments recommending readability of `= value` and still obtaining efficiency under the hood.

I'm bringing that up more as the battles that projects face with the langauge itself, and how the C++ committee shoots itself in the foot by making these constructs that introduce divisiveness in projects, and how we're all a victim to it here. 

So yeah: here were are claiming to using modern languages and practices.. and yet we're not drinking _all_ of the C++17 and soon-to-be C++20 coolaid.

I'm sure we're in for more syntax A vs. syntax B vs. syntax C battles.. and yeah, if we don't keep pace, maybe someone will fork Staging itself _"A fork that welcomes all C++20 (and beyond) syntax to the maximum." 

So we need to keep pace.

I feel like this last PR, he really tested the waters by using that lame var{}; initializer nonsense..  he mentioned how he had previously decided it was going to be his last PR, so I feel like this was kind of deliberate.

I still feel like there's no problem with using `std::move` and `string_view` within your PR: however, it simply needs to be justified with some comments, like "Move the item from this vector to that vector" -> and it would be self-documenting, and to address the comment that this is more about object ownership and not about performance (to address @johnnovak's concern).

There's also the big picture: this might be the only C++ implementation of DOS `FOR` handling out there.  That old patch on vogons was using lots of C string manipulations, so given the balance: I would prefer yours straight away.

Likewise: dig up FreeDOS's `FOR` implementation (which I can't find), but I guarantee isn't highly non-trivial and very low level.  Similarly if you have access to the original MS DOS implementation: that's going to be some extremely involved code. 

So yeah: in the end, `FOR` is an addon that doesn't tough the normal flow, so it's an incremental improvement we want.

Reply:
To be fair, just about every project has rules about which parts of C++ to use because the language is huge and some features were designed by the drunk and/or mentally deranged.  "C++ to the max" sounds like "every little thing has to be a templated class that fits into our system-wide multiple inheritance hierarchy and if we ever catch you with a raw pointer we ban you from the project".

You kind of have to pick some subset of C++ in order to stay sane and I've seen it range from "use almost exclusively C style code" to "classes are fine but don't use the STL" to "use modern features but stay away from the crazy bits" to "go nuts and good luck figuring this shit out" and anywhere in between.

Which type of initializer to use is almost strictly stylistic btw.  I agree it's important to stay consistent though or code gets hard to read.  Same with indentation style, snake_case vs MicrosoftCase, etc.  I don't care that much what the style is for a given project, so long as it's consistent.

>  On the other hand, compilers are free to eliminate unnecessary temporaries: so we see counter-arguments recommending readability of = value and still obtaining equivalent efficiency under the hood.

Prior to C++ 17, this is true.  Now, compilers are _required_ to eliminate the temporary so like I said strictly stylistic (and also annoying that there's two forms of syntax that do exactly the same thing).

https://en.cppreference.com/w/cpp/language/copy_elision

>  Since C++17, a prvalue is not materialized until needed, and then it is constructed directly into the storage of its final destination. This sometimes means that even when the language syntax visually suggests a copy/move (e.g. [copy initialization](https://en.cppreference.com/w/cpp/language/copy_initialization)), no copy/move is performed — which means the type need not have an accessible copy/move constructor at all.

Reply:
I like STL containers, they do the job. We also started using `optional` which is great, and I'm also experimenting with `variant` and `visit` in my upcoming DOS command PR as we speak. Overall, I like these two features more than I dislike they slight verboseness because they simplify the code (I think) and they could establish a common pattern for writing unit testable DOS programs that have lots of command line switches.

But I will ask the opinion of the team about their use: if everybody hates them, I will try to find an alternative solution. Ironically, I've also started experimenting with the inline creation of structs with braces _sparingly_, only where it increases readability and cuts down on the temporaries and the number of lines.

I'm just saying this because it's not like I'm completely against the usage of _any_ new C++ features, but I definitely want to do it sparingly, with approval from the team, and considering every single new feature individually and carefully.

I can be very easy to work with, or very annoying to work with (I guess) because I have to be convinced about the overarching, global benefits of introducing something new before I'm happy with it. That includes readability, maintainability, comfort level of the team, my own comfort level, is this feature _fun_ to use, performance, how will this increase or decrease the approachability of the project for others, and the list goes on. If overall I think the sum of all these parts is not positive, I don't care if the thing in question is the most trendy thing to do, or the "best practice", or was blessed by committe X, or is the favourite pet feature of the person I'm talking with, or brings 10-20% benefit in a few items on the list of considerations, but has otherwise a net negative effect.

This FOR implementation is surely an improvement over existing alternative implementations @MeAreJeenius. If you would still like to finish the job, I think the next highest value thing would be to put a comprehensive unit test suite around it. We can keep the string views and the move, or remove them later; all in all, and big picture again, I care a lot more whether the feature is functionally sound and compatible with DOS 6.22. So if you would do that next, that would be much appreciated and.